src/merge_maps_kinetic.cpp

```diff
@@ -166,51 +183,97 @@ bool MergeMapTool::MergeMapCallback(slam_toolbox::MergeMaps::Request &req,
                                     slam_toolbox::MergeMaps::Response &resp)
 /*****************************************************************************/
 {
-  //TODO
-  // take submaps and project all positions into map frame,submap_locations_
-  // scans
-
-  // create map using kartooccupany grid from all in same frame
-  // KartoToROSOccupancyGrid(scans)
-
-  // publish it for visualization/map_saver using sstMS_[0]'s for meta and normal'
-
-  // later, take nodes in graph mutually closest to each other and scana match
-  // against them to make a new inter-graph contraint list to optimize over
-  // then generate new composite map
-  // (this same technique can then be applied with manual loop closures)
+    int id = 0;
+    karto::Pose2 corrected_pose;
+    karto::LocalizedRangeScan* pScan;
+    karto::LocalizedRangeScanVector transformed_scans;
+    std::vector<karto::LocalizedRangeScanVector>& vector_of_scans = scans_vec_;
+    karto::LocalizedRangeScan* pScan_copy;
+
+    for(LocalizedRangeScansVec_it it_LRV = vector_of_scans.begin(); it_LRV!= vector_of_scans.end(); ++it_LRV)
+    {
+      id++;
+      for ( LocalizedRangeScans_it iter = (*it_LRV).begin(); iter != (*it_LRV).end();++iter)
+      {
+        pScan= *iter;
+        pScan_copy = pScan;
+        tf::Transform submap_correction = submap_marker_transform_[id];
+
+        //TRANSFORM BARYCENTERR POSE
+        const karto::Pose2 baryCenter_pose = pScan_copy->GetBarycenterPose();
+        Pose2_ptr karto_baryCenterPose_corr = std::move(ApplyCorrection(baryCenter_pose, submap_correction));
+        pScan_copy->SetBarycenterPose(*karto_baryCenterPose_corr);
+
+        //TRANSFORM BOUNDING BOX POSITIONS
+        karto::BoundingBox2 bbox = pScan_copy->GetBoundingBox();
+        const Vector2_double_ptr bbox_min_corr = std::move(ApplyCorrection(bbox.GetMinimum(), submap_correction));
+        bbox.SetMinimum(*bbox_min_corr);
+        const Vector2_double_ptr bbox_max_corr =std::move(ApplyCorrection(bbox.GetMaximum(), submap_correction));
+        bbox.SetMaximum(*bbox_max_corr);
+        pScan_copy->SetBoundingBox(bbox);
+
+        // TRANSFORM UNFILTERED POINTS USED
+        karto::PointVectorDouble UPR_vec = pScan_copy->GetPointReadings();
+        for(karto::PointVectorDouble::iterator it_upr = UPR_vec.begin(); it_upr!=UPR_vec.end(); ++it_upr)
+        {
+          const Vector2_double_ptr upr_corr = std::move(ApplyCorrection(*it_upr, submap_correction));
+          (*it_upr).SetX(upr_corr->GetX());
+          (*it_upr).SetY(upr_corr->GetY());
+        }
+        pScan_copy->SetPointReadings(UPR_vec);
+
+        //TRANSFORM CORRECTED POSE
+        corrected_pose = pScan_copy->GetCorrectedPose();
+        Pose2_ptr karto_robotPose_corr = std::move(ApplyCorrection(corrected_pose, submap_correction));
```

Comment:
You dont need a pointer and in fact makes this much harder. Remove all the pointers from these corrections.